zio-http/shared/src/main/scala/zio/http/Server.scala

```diff
@@ -462,15 +476,24 @@ object Server extends ServerPlatformSpecific {
 
   private final case class ServerLive(
     driver: Driver,
-    bindPort: Int,
+    // A promise used to signal the first time `install`
+    // is called on this `Server` instance.
+    private val initialInstall: Promise[Nothing, Unit],
+    // A promise that represents the port of the "started" driver
+    // or a throwable if starting the driver failed for any reason.
+    private val serverStarted: Promise[Throwable, Int],
   ) extends Server {
     override def install[R](httpApp: Routes[R, Response])(implicit
       trace: Trace,
       tag: EnvironmentTag[R],
     ): URIO[R, Unit] =
-      ZIO.environment[R].flatMap(env => driver.addApp(httpApp, env.prune[R]))
+      for {
+        _ <- initialInstall.succeed(())
+        _ <- serverStarted.await.orDie
+        _ <- ZIO.environment[R].flatMap(env => driver.addApp(httpApp, env.prune[R]))
+      } yield ()
 
-    override def port: Int = bindPort
+    override def port[R]: URIO[R, Int] = serverStarted.await.orDie
```

Comment:
`serverStarted` will block indefinitely until `install` is called for the first time. Should we add a comment that mentions this?

Reply:
Also, I don't love that this now returns a `ZIO` -- I couldn't think of any way around this since port may or may not be present depending on if the driver has been started.